File: scripts/validate-labs.mjs
// Validates every lab in the same Pyodide runtime and harness the browser
// uses (public/nl_harness.py):
//   - code steps: the starter must NOT pass all checks; the solution must pass all
//   - predict steps: the marked answer must match what the code really prints
//   - explain steps: the model answer must cover every key idea
// Run with `npm run validate:labs`. Exits non-zero on any failure.

import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import path from "node:path";

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const { loadPyodide } = await import(path.join(root, "public/pyodide/pyodide.mjs"));
const labs = JSON.parse(readFileSync(path.join(root, ".labs-build/labs.json"), "utf8"));
const only = process.argv[2];

let out = [];
const py = await loadPyodide({
  indexURL: path.join(root, "public/pyodide/"),
  packageBaseUrl: "https://cdn.jsdelivr.net/pyodide/v314.0.7/full/",
  stdout: (s) => out.push(s),
  stderr: () => {},
});
py.runPython(readFileSync(path.join(root, "public/nl_harness.py"), "utf8"));
const prepared = new Set();

async function prepare(packages = []) {
  const missing = packages.filter((p) => !prepared.has(p));
  if (!missing.length) return;
  await py.loadPackage(missing, { messageCallback: () => {} });
  py.globals.get("_nl_prepare")(py.toPy(missing));
  missing.forEach((p) => prepared.add(p));
}

function run(code, files) {
  for (const [name, content] of Object.entries(files ?? {})) py.FS.writeFile(name, content);
  out = [];
  const ns = py.toPy({});
  const err = py.globals.get("_nl_run")(code, ns);
  py.globals.get("_nl_figures")(ns);
  ns.set("_stdout", out.join("\n"));
  ns.set("_source", code);
  const error = err ? err.toJs({ dict_converter: Object.fromEntries }) : null;
  return { ns, error, stdout: out.join("\n") };
}

function check(exprs, ns) {
  return py.globals.get("_nl_check")(py.toPy(exprs), ns).toJs();
}

let bad = 0;
const report = (ok, tag, detail = "") => {
  if (!ok) bad++;
  console.log(`${ok ? "OK " : "BAD"} ${tag}${ok ? "" : "  " + detail}`);
};

for (const lab of labs) {
  if (only && lab.slug !== only) continue;
  await prepare(lab.packages);
  for (const s of lab.steps) {
    const tag = `${lab.slug}/${s.id}`;
    if (s.kind === "predict") {
      const r = run(s.code, lab.files);
      const got = r.error ? r.error.summary : r.stdout;
      const want = s.options[s.answer];
      report(got === want, tag, `got ${JSON.stringify(got)} want ${JSON.stringify(want)}`);
    } else if (s.kind === "code") {
      const exprs = s.checks.map((c) => c.expr);
      const st = run(s.starterCode, lab.files);
      const r0 = st.error ? exprs.map(() => false) : check(exprs, st.ns);
      if (!s.solution) {
        report(false, tag, "no solution");
        continue;
      }
      const so = run(s.solution, lab.files);
      const r1 = so.error ? exprs.map(() => false) : check(exprs, so.ns);
      const ok = r1.every(Boolean) && !r0.every(Boolean);
      report(ok, tag, `starter=${JSON.stringify(r0)} solution=${JSON.stringify(r1)} solutionError=${so.error?.summary ?? "none"}`);
    } else if (s.kind === "explain") {
      const covered = s.ideas.map((i) => i.patterns.some((p) => new RegExp(p, "i").test(s.modelAnswer)));
      report(covered.every(Boolean), tag, `model answer misses ${JSON.stringify(s.ideas.filter((_, i) => !covered[i]).map((i) => i.label))}`);
    }
  }
}

console.log(`\n${bad} failure${bad === 1 ? "" : "s"}`);
process.exit(bad ? 1 : 0);

File: scripts/validate_labs.py
"""Validates every lab: starter code must fail its checks, the solution must
pass them, predictions must match real output, and each reflection's model
answer must cover its key ideas. Mirrors the checks harness in
public/pyodide-worker.js. Run with `npm run validate:labs`."""
import json, os, sys, io, ast, contextlib, tempfile, traceback, re
labs = json.load(open(sys.argv[1]))

def run(code, files):
    d = tempfile.mkdtemp(); os.chdir(d)
    for n, c in (files or {}).items(): open(n, "w").write(c)
    ns = {}; out = io.StringIO(); err = None
    try:
        with contextlib.redirect_stdout(out):
            exec(compile(code, "main.py", "exec"), ns)
    except BaseException as e:
        err = "".join(traceback.format_exception_only(e)).strip().splitlines()[-1]
    ns["_stdout"] = out.getvalue().rstrip("\n"); ns["_source"] = code
    def _with(**ov):
        tree = ast.parse(code); ins = {}
        for k, v in ov.items():
            for node in tree.body:
                if isinstance(node, ast.Assign) and any(isinstance(t, ast.Name) and t.id == k for t in node.targets):
                    ins.setdefault(node.end_lineno, []).append(f"{k} = {v!r}"); break
        o = []
        for i, l in enumerate(code.split("\n"), 1): o.append(l); o.extend(ins.get(i, []))
        n2 = {}
        with contextlib.redirect_stdout(io.StringIO()): exec(compile("\n".join(o), "main.py", "exec"), n2)
        return n2
    ns["_with"] = _with
    return ns, err

def checks(step, ns):
    res = []
    for c in step["checks"]:
        try: res.append(bool(eval(c["expr"], ns)))
        except Exception as e: res.append(False)
    return res

bad = 0
for lab in labs:
    for s in lab["steps"]:
        tag = f'{lab["slug"]}/{s["id"]}'
        if s["kind"] == "predict":
            ns, err = run(s["code"], lab.get("files"))
            got = ns["_stdout"] if not err else err
            want = s["options"][s["answer"]]
            ok = got == want
            print(("OK  " if ok else "BAD ") + tag, "" if ok else f"got {got!r} want {want!r}"); bad += not ok
        if s["kind"] == "code":
            ns, err = run(s["starterCode"], lab.get("files"))
            r0 = checks(s, ns) if not err else [False]*len(s["checks"])
            sol = s.get("solution")
            if not sol:
                print("NOSOL", tag); continue
            ns, err2 = run(sol, lab.get("files"))
            r1 = checks(s, ns)
            ok = all(r1) and not all(r0) and not err2
            print(("OK  " if ok else "BAD ") + tag, f"starter_err={err} starter={r0} solution={r1} solerr={err2}"); bad += not ok
        if s["kind"] == "explain":
            m = s["modelAnswer"]
            cov = [any(re.search(p, m, re.I) for p in i["patterns"]) for i in s["ideas"]]
            ok = all(cov)
            print(("OK  " if ok else "BAD ") + tag, "model answer covers", cov); bad += not ok
print("BAD:", bad)
sys.exit(1 if bad else 0)
